### PR TITLE
proxy: Parse units with duration configurations

### PR DESCRIPTION
**Pull request by olix0r**
Configuration values that take durations are currently specified as
time values with no units.  So `600` may mean 600ms in some contexts and
10 minutes in others.

In order to avoid this problem, this change now requires that
configurations provide explicit units for time values such as '600ms' or
'10 minutes'.

After this, users may set environment vars like `CONDUIT_PROXY_METRICS_RETAIN_IDLE="1h"`.

### CODE REVIEW

File: proxy/src/config.rs [hunk 1]
```diff
@@ -340,6 +339,42 @@ fn parse_number<T>(s: &str) -> Result<T, ParseError> where T: FromStr {
     s.parse().map_err(|_| ParseError::NotANumber)
 }
 
+fn parse_duration(s: &str) -> Result<Duration, ParseError> {
+    use regex::Regex;
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"(?x)^
+            \s* (\d+) \s*
+            ( ms | msecs? | millis?  | milliseconds?
```
Comment (briansmith) [inline proxy/src/config.rs]: I think we should only support the short form of each duration unit that we support: "ms", "s", "m", "h", "d", to make life easier and to make sure our testing is comprehensive.

File: proxy/src/config.rs [hunk 2]
```diff
@@ -340,6 +339,42 @@ fn parse_number<T>(s: &str) -> Result<T, ParseError> where T: FromStr {
...
+        return Ok(Duration::from_millis(magnitude));
+    }
+
+    let to_secs = match unit.chars().next() {
+        Some('s') => 1,
+        Some('m') => 60,
+        Some('h') => 60 * 60,
+        Some('d') => 60 * 60 * 24,
+        _ => return Err(ParseError::NotADuration),
+    };
```
Comment (briansmith) [inline proxy/src/config.rs]: In particular, notice that this doesn't validate anything after the first character, which I'm sure will create problems later, if we change anything here later.

Instead, we should match `unit` as full strings. Then we can also merge the processing for "ms" into the match and use `from_millis` for everything.

File: proxy/src/config.rs [hunk 3]
```diff
@@ -380,3 +415,65 @@ fn parse<T, Parse>(strings: &Strings, name: &str, parse: Parse) -> Result<Option
         None => Ok(None),
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn test_units<F: Fn(u64) -> Duration>(units: &[&str], to_duration: F) {
+        for v in &[1, 23, 456_789] {
```
Comment (briansmith) [inline proxy/src/config.rs]: We should include 0 here too. We should consider whether we support the value 0 at all for any of these values, and, if so, whether we require a unit for 0.

File: proxy/src/config.rs [hunk 4]
```diff
@@ -380,3 +415,65 @@ fn parse<T, Parse>(strings: &Strings, name: &str, parse: Parse) -> Result<Option
...
+    #[test]
+    fn parse_float_numbers_are_invalid() {
+        assert_eq!(parse_duration(".123 hours"), Err(ParseError::NotADuration));
+        assert_eq!(parse_duration("1.23 hours"), Err(ParseError::NotADuration));
+    }
+
+    #[test]
+    fn parse_invalid_unit() {
+        assert_eq!(parse_duration("12 moons"), Err(ParseError::NotADuration));
+    }
```
Comment (briansmith) [inline proxy/src/config.rs]: We should also have a test for the input string that would cause a overflow, as well as for the maximum value that won't cause an overflow. Also negative numbers.

**Comment by briansmith** [comment]
Note that this fixes #27.

**Comment by briansmith** [comment]
I imagine `conduit inject` and `conduit install` need to be modified to add the units to the config files they generate, and to accept the units in their command-line argument parsing.

**Comment by olix0r** [comment]
@briansmith these values are not currently configurable in the CLI, so no change is necessary

**Comment by briansmith** [comment]
OK, I thought #865 was merged already. #865 adds one.

**Comment by olix0r** [comment]
@briansmith i've removed support for all but the shortest names and i've added tests for "0" and an overflow number
